### ISSUE TITLE
Controller 'uibTabset', required by directive 'uibTabContentTransclude', can't be found!

### ISSUE THREAD

**Issue by spfiedler** [angular-ui/bootstrap#5350]
This occurred when I upgraded from 0.14 to 1.1.

I have a custom directive that displays uib-tab elements in two places.  The template for 'my-tab-container' directive is:

&lt;uib-tabset my-tab-container1 justified="false" type="pills"&gt;
    &lt;ng-transclude &gt;&lt;/ng-transclude&gt;
&lt;/uib-tabset&gt;

&lt;uib-tabset ng-if="! omitLowerTabs" my-tab-container2 justified="false" type="pills"&gt;
    &lt;ng-transclude &gt;&lt;/ng-transclude&gt;
&lt;/uib-tabset&gt;

I call the directive with the following test:

    var element = createElement('<my-tab-container>' +
        '<uib-tab vc-tab heading="Margin"><div style="height:100px; margin:10px">Content with height and margin on div</div></uib-tab>' +
        '<uib-tab vc-tab heading="Height"><div style="height:100px">Content with height but no margin on the div</div></uib-tab>' +
        '<uib-tab vc-tab heading="Very Short"><div>Content with no height or margin on div</div></uib-tab>' +
        '</my-tab-container>');

See how I transclude in the uib-tab elements into the template.
However, when I run the test I get the error as stated in the subject line.  The browser also logs the error in the console, however the behavior is still as expected.

-----
I researched the uibTabContentTransclude directive.  

.directive('uibTabContentTransclude', function() {
  return {
    restrict: 'A',
    require: '^uibTabset',
    link: function(scope, elm, attrs) {
      var tab = scope.$eval(attrs.uibTabContentTransclude);
      .....

When I make a change the the 'require' attribute to make it optional, the content transclusion occurs correctly, e.g.

    require: '^?uibTabset',

-----
Can the require specification for this directive be relaxed so this type of transclusion can work?

**Comment by wesleycho** [comment]
Looking at the code, would the linking function correctly execute still and the tabs function as they should?

I'm concerned about the effect this would have on the linking function arguments...this smells of poor API design of Angular if this is how it behaves.

**Comment by spfiedler** [comment]
Yes, the linking function correctly execute and the tabs function as they should.  

To be clear, this problem occurrs with:
    "angular": "~1.4.0",
    "angular-bootstrap": "~1.1.0",

Previously when I used
    "angular": "~1.4.0",
    "angular-bootstrap": "~0.14.0",

everything was transcluded correctly and it worked great.

**Comment by spfiedler** [comment]
I updated the example above because it was not showing the main part of the problem.  Notice ng-transclude is performed in two different parts of the template.  This is certainly legal, however, the updated uib-tabs directive comes up with the error, whereas it did not in the previous release.  The minor change to the the 'require' attribute fixes all.

**Comment by mrzepinski** [comment]
It's related to https://github.com/angular-ui/bootstrap/issues/3692
I have to same problem. When I change $state I get an error.

**Comment by spfiedler** [comment]
@wesleycho - is there any way to get this minor 1 character change in?  I know I have had to add the '?' to various arguments in my angular directives because it was not guaranteed that the element would be present initially.  I have tested it and find no ill effects.  Otherwise I will have to probably perform a BIG workaround.  And as I said, this WORKED in a previous release.  It's only because those required clauses were added everywhere that it became an issue.  Thx.

**Comment by wesleycho** [comment]
This seems like a dangerous change due to how Angular works, and I'm quite wary about directives that try to contain one portion of highly coupled parent-child directives and not the complete unit. I'm inclined to reject this request, but leaving open for some more comment first.

Would it not be possible to do something like this
```html
<my-tab-container>
  <uib-tabset>
    <uib-tab></uib-tab>
  </uib-tabset>
</my-tab-container>
```

The problem with the original request as is is that this sort of abstraction is extremely fragile with Angular (and perhaps any UI library), as it separates tightly coupled & inter-dependent directives, which then can create larger problems for the maintenance of them because it loosens control of what is at its root tightly coupled API. That Angular throws errors here when it comes to required controllers is no mistake here IMO - it likely is trying to discourage this sort of usage.

**Comment by icfantv** [comment]
I've been meaning to create a working example of tabs w/ nested controllers - it's still on my radar.  That said, I thought we decided that the right way to solve this was to use directives INSIDE the `uib-tab` elements.

I would tend to agree with @wesleycho in that your proposed solution of optionally decoupling the two elements we designed to be coupled is wrong.  Additionally, the transclusion of the tabs into the tabset is really weird and feels like a bad design.

I can certainly appreciate that rearchitecting things to make them "correct" sucks and can be a huge time sink and that "just making this one change" makes your life insanely easier.  Sometimes, we as devs, just need to bite the bullet and do the refactor we've been dreading.
